src/interfaces/protocols-configure.ts

```diff
@@ -79,24 +79,42 @@ export class ProtocolsConfigure extends Message<ProtocolsConfigureMessage> {
     // Traverse nested rule sets
     for (const rootRecordPath in definition.structure) {
       const rootRuleSet = definition.structure[rootRecordPath];
-      ProtocolsConfigure.validateRuleSet(rootRuleSet, rootRecordPath, globalRoles);
+
+      // gather $contextRoles
+      const contextRoles: string[] = [];
+      for (const childRecordType in rootRuleSet) {
+        if (childRecordType.startsWith('$')) {
+          continue;
+        }
+        const childRuleSet: ProtocolRuleSet = rootRuleSet[childRecordType];
+        if (childRuleSet.$contextRole) {
+          contextRoles.push(`${rootRecordPath}/${childRecordType}`);
+        }
+      }
+
+      ProtocolsConfigure.validateRuleSet(rootRuleSet, rootRecordPath, [...globalRoles, ...contextRoles]);
```

Comment:
Sanity: the placement of this line and the repeated concatenation of global and context roles threw me off: doesn't this mean that the `contextRoles` won't be built fully until the end of outer for-loop, leading to bugs in `validateRuleSet()`?

Consider moving this line into its own loop after the full `contextRoles` is built. I know it's literally repeating the same looping boilerplate 3 times, but I really liked it when you did it 2 times originally to reduce cognitive load, making it easy to comprehend and less error-prone!

Reply:
We need to construct the `contextRoles` separately for each root record type. The list of `contextRoles` isn't unfinished, it's deliberately different for each iteration of the `for` loop.

Consider this protocol definition structure that has both `threads` and `blogs`, with context roles `thread/participant` and `blog/subscriber`. In the `thread` tree protocol rules, there should never be an `$action` rule containing `blog/subscriber` because there can never be a `blog/subscriber` in the same context as a `thread/chat`.

```javascript
{
  thread: {
     participant: {
       $contextRole: true,
     },
     chat: {
       $actions: [{
         role: 'thread/participant',
         can: 'read'
       },
       {
         // should not be allowed since 'blog/subscriber' roles can never exist in this context
         role: 'blog/subscriber',
         can: 'read'
       }]
     }
  },
  blog: {
    $actions: [{
      role: "blog/subscriber",
      can: "read"
    }]
    subscriber: {
      $contextRole: true
    }
  }
}
```

Reply:
I see! Thanks, makes sense now!